diff --git a/src/tools/scamv/bir_embexp_driverLib.sig b/src/tools/scamv/bir_embexp_driverLib.sig
index 561a77a..56a8b5e 100644
--- a/src/tools/scamv/bir_embexp_driverLib.sig
+++ b/src/tools/scamv/bir_embexp_driverLib.sig
@@ -64,7 +64,7 @@ signature bir_embexp_driverLib = sig
 
   (* Input: exp_id *)
   (* Output: asm_lines, model pair *)
-  val bir_embexp_load_exp  : string -> string list * (modelValues list * modelValues list)
+  val bir_embexp_load_exp  : string -> string list * (modelValues list * modelValues list) * modelValues list option
 
   val bir_embexp_load_exp_ids : string -> string list
 
diff --git a/src/tools/scamv/bir_embexp_driverLib.sml b/src/tools/scamv/bir_embexp_driverLib.sml
index bb6fa1e..0d77aea 100644
--- a/src/tools/scamv/bir_embexp_driverLib.sml
+++ b/src/tools/scamv/bir_embexp_driverLib.sml
@@ -592,10 +592,16 @@ end
       val input1_file = (logs_dir ^ "/" ^ exp_id ^ "/input1.json");
       val input2_file = (logs_dir ^ "/" ^ exp_id ^ "/input2.json");
 
-      val s = (parse_back_json_state false input1_file,
-               parse_back_json_state false input2_file);
+      val train_file = (logs_dir ^ "/" ^ exp_id ^ "/train.json");
+      val train_filename = (read_from_file train_file; SOME train_file)
+                           handle _ => NONE;
+
+      val s1 = parse_back_json_state false input1_file;
+      val s2 = parse_back_json_state false input2_file;
+
+      val traino = Option.map (fn filename => parse_back_json_state false filename) train_filename;
     in
-      (asm_lines, s)
+      (asm_lines, (s1,s2), traino)
     end;
 
   fun bir_embexp_load_list listtype listname =
diff --git a/src/tools/scamv/examples/analyzeprogs.sml b/src/tools/scamv/examples/analyzeprogs.sml
index 8c4c639..46b6d18 100644
--- a/src/tools/scamv/examples/analyzeprogs.sml
+++ b/src/tools/scamv/examples/analyzeprogs.sml
@@ -38,20 +38,26 @@ fun get_birexp_vars exp =
 
 (* branch: tacas_29_09_01_cache_addr_pc *)
 val listname = "M1_zeroedstate_160";
+val (arch_id, exp_type_id, state_gen_id) =
+    ("arm8", "cache_multiw", "regen_for_hypothesis");
+val prog_gen_id = "regen_for_hypothesis_this_prog_should_exist_already";
 
 val exp_ids = bir_embexp_load_exp_ids listname;
 val exp_id = List.nth(exp_ids,1);
 (*
 val exp_id = "arm8/exps2/cache_multiw/96092ba1f752f2e06ffefb9a186575c6351bf9ab";
 *)
+val exp_id = List.last exp_ids;
 
-val _ = List.map (fn exp_id =>
+val exp_ids_new = List.map (fn exp_id =>
   let
 
 val _ = print ("\n=========================================================\n");
 val _ = print (exp_id ^ "\n");
 
-val (asm_lines, s) = bir_embexp_load_exp exp_id;
+val (asm_lines, (s1,s2), traino) = bir_embexp_load_exp exp_id;
+val s_train = valOf traino
+              handle _ => raise ERR "script" "no training data";
 
 (*
 val asm_code = bir_embexp_prog_to_code asm_lines;
@@ -140,7 +146,6 @@ val pathcondvarsmap = List.map pathtocondvars paths_i;
 val _ = if length pathcondvarsmap = 2 then () else
         raise ERR "script" "was checked before";
 
-val (s1,s2) = s;
 (*
 val s_ = s2;
 *)
@@ -371,6 +376,22 @@ val _ = print ("shrinked s1 from " ^ (Int.toString (get_state_size s1)) ^ " to "
 val _ = print ("shrinked s2 from " ^ (Int.toString (get_state_size s2)) ^ " to " ^ (Int.toString (get_state_size s2_filtered)) ^ "\n");
 val _ = print ("\n");
 
+val code_asm = bir_embexp_prog_to_code asm_lines;
+val prog_id = bir_embexp_prog_create (arch_id, prog_gen_id) code_asm;
+
+val exp_id_new = bir_embexp_sates3_create (arch_id, exp_type_id, state_gen_id) prog_id (s1,s2,s_train);
+
   in
-    ()
+    exp_id_new
   end) exp_ids;
+
+
+val _ = print ("=================================================");
+val _ = print (":::::::::::::::::::::::::::::::::::::::::::::::::");
+val _ = print ("=================================================");
+
+val _ = List.map (fn exp_id =>
+  let
+    val _ = print (exp_id ^ "\n");
+  in () end
+) exp_ids_new;
