diff --git a/src/tools/scamv/examples/analyzeprogs.sml b/src/tools/scamv/examples/analyzeprogs.sml
index 44dd25e..c7b8b58 100644
--- a/src/tools/scamv/examples/analyzeprogs.sml
+++ b/src/tools/scamv/examples/analyzeprogs.sml
@@ -1,6 +1,6 @@
 
 (* branch: tacas_29_09_01_cache_addr_pc *)
-val listname = "test123";
+val listname = "tacas_cache_addr_pc_p1_0";
 
 (*
 =================================== general stuff starts here ====================================
